File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        Deque<Pair<TreeNode, Integer>> stack = new ArrayDeque<>();
        int res = 0;
        stack.push(new Pair(root, 1));

        while(!stack.isEmpty()){
            Pair<TreeNode,Integer> current = stack.pop();
            TreeNode node = current.getKey();
            int depth = current.getValue();

            if(node != null){
                res = Math.max(res, depth);
                if(node.left != null) stack.push(new Pair(node.left, depth+1));
                if(node.right != null) stack.push(new Pair(node.right, depth+1));
            }
        }
        return res;
    }
}

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        def dfs(node, depth) -> int:
            if not node:
                return 0

            left = dfs(node.left, depth)
            right = dfs(node.right, depth)
            
            depth = 1 + max(left, right)

            return depth

        return dfs(root, 0)
        